### ISSUE TITLE
RadioTile Component causing accessibility failures.

### ISSUE THREAD

**Issue by AsthaJain1** [IBM/carbon-components-react#1114]
## Detailed description

> Describe in detail the issue you're having.
I am using Selectable Tile component. 
http://react.carbondesignsystem.com/?selectedKind=Tile&selectedStory=Selectable&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

> Is this a feature request (new component, new icon), a bug, or a general issue?

Bug - causing accessibility failures at my end.
> Is this issue related to a specific component?

http://react.carbondesignsystem.com/?selectedKind=Tile&selectedStory=Selectable&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Factions%2Factions-panel

> What did you expect to happen? What happened instead? What would you like to see changed?
Description of accessibility failure - 
https://aat.mybluemix.net/token/97b22e9c-25bd-48ea-be04-c77a42c72b11/a313b75d-5599-420e-9695-15c1599da92a/doc/w3/help/en-US/idhi_accessibility_check_g1062.html

The input has a css of `display: none` causing this failure.

> What browser are you working in?
Chrome
> What version of the Carbon Design System are you using?
    "carbon-components": "^8.18.18",
    "carbon-components-react": "^5.52.4",
> What offering/product do you work on? Any pressing ship or release dates we should be aware of?
IBM Cloud

**Comment by ZachStowellIBM** [comment]
@alisonjoseph Our project is still seeing this a11y issue in chrome, using "carbon-components": "^9.18.0" and  "carbon-components-react": "^6.62.2".

I've found a hacky workaround for this initial a11y error, however it leads to more issues that would require some major changes to the RadioTile component, as well as possibly some changes to the TileGroup component (and possibly changes in the `carbon-components` library).

Looking at https://xo33l4xqxp.codesandbox.io/ ([editable](https://codesandbox.io/s/xo33l4xqxp)), I'm seeing the same a11y failure as above.

When I modify the input to not use `display:none` as shown in https://l32kxkv3nm.codesandbox.io/ ([editable](https://codesandbox.io/s/l32kxkv3nm)), it introduces additional a11y errors https://aat.mybluemix.net/token/97b22e9c-25bd-48ea-be04-c77a42c72b11/a313b75d-5599-420e-9695-15c1599da92a/doc/w3/help/en-US/idhi_accessibility_check_g1018.html and https://aat.mybluemix.net/token/97b22e9c-25bd-48ea-be04-c77a42c72b11/a313b75d-5599-420e-9695-15c1599da92a/doc/w3/help/en-US/idhi_accessibility_check_g1029.html

I found https://github.com/IBM/carbon-components-react/pull/1691 that @emyarod worked on, and it seems to address other a11y issues, however it doesn't cover this specific issue.  Additionally, I found [this issue](https://github.com/IBM/carbon-components/issues/1522) on the carbon-components repo -- would it be reasonable to expect that this a11y failure would be updated when v10 of carbon-components is released?

**Comment by dakahn** [comment]
Good morning. The a11y issue you're experiencing is not a bug.

When you use `display: none` to hide an element from your users you're also removing it from the document object model. The document object model is how tools like screen readers understand your web content.

So your label is pointing to an id that doesn't exist on the page (since it has been removed from the DOM). Thus the error.

If you need to hide this element visually and have it appear in the document object model there are a few different ways of doing this.

Here's a couple links elaborating on why we should avoid `display: none` for hiding elements and alternative techniques to use instead:

https://alistapart.com/article/now-you-see-me

https://css-tricks.com/places-its-tempting-to-use-display-none-but-dont/

**Comment by ZachStowellIBM** [comment]
@dakahn Thanks for the comment.  The `display:none` is currently in the `<RadioTile>` component, which is why the original issue was raised -- the `<RadioTile>` doesn't pass certain a11y requirements.

![screen shot 2019-01-09 at 10 27 10 am](https://user-images.githubusercontent.com/39738800/50909255-790a1880-13f9-11e9-964e-5bbf65a40283.png)
![screen shot 2019-01-09 at 10 28 20 am](https://user-images.githubusercontent.com/39738800/50909268-80c9bd00-13f9-11e9-819d-2a17e35d721e.png)

When I removed the `display:none` with a different method of hiding the input ([code sandbox link](https://codesandbox.io/s/l32kxkv3nm)), it produces new a11y issues due to the way the `<RadioTile>` markup is constructed.

After the `display:none` is addressed, the two new a11y issues are:
 - [Check boxes and radio buttons must have a label after the input control](https://aat.mybluemix.net/token/97b22e9c-25bd-48ea-be04-c77a42c72b11/a313b75d-5599-420e-9695-15c1599da92a/doc/w3/help/en-US/idhi_accessibility_check_g1018.html)
 - [Groups of radio buttons or check boxes with the same name attribute should be contained within a fieldset element](https://aat.mybluemix.net/token/97b22e9c-25bd-48ea-be04-c77a42c72b11/a313b75d-5599-420e-9695-15c1599da92a/doc/w3/help/en-US/idhi_accessibility_check_g1029.html)

Addressing those new issues would be a major change to the `<RadioTile>` component, possibly the `<TileGroup>` component (for the fieldset issue), as well the CSS changes.

@emyarod I see you assigned yourself to this issue -- is there any additional info or anything that would be useful to help?

**Comment by emyarod** [comment]
hi @AsthaJain1 and @ZachStowellIBM thanks for reporting this issue! Just to clarify, the original issue is caused by the lack of a `for` attribute on the `label` for our `<Tile>` components is that correct?

**Comment by ZachStowellIBM** [comment]
@emyarod the issue is actually that the <input> that the `for` references has a `display:none` style for the `<RadioTile>` component (selectable tile), which makes the reference invalid.  

Fixing the `display:none` by using any of the techniques mentioned in @dakahn 's comment, or by the one I did in [this code sandbox](https://codesandbox.io/s/l32kxkv3nm) introduces new a11y issues, which I've included in the screenshots below.

_a11y issues_
![screen shot 2019-01-09 at 11 31 48 am](https://user-images.githubusercontent.com/39738800/50913680-dbb3e200-1402-11e9-8ee6-46c947486490.png)

_CSS Changes_
![screen shot 2019-01-09 at 11 35 00 am](https://user-images.githubusercontent.com/39738800/50913717-f25a3900-1402-11e9-8057-8f85873ab426.png)

**Comment by ZachStowellIBM** [comment]
Are there any updates on this fix?  I see that https://github.com/IBM/carbon-components-react/pull/1774 is still awaiting an additional review.
